implementations/rust/ockam/ockam/src/metadata.rs

```diff
@@ -0,0 +1,159 @@
+use core::ops::{Deref, DerefMut};
+use ockam_core::{
+    compat::{collections::BTreeMap, string::String, vec::Vec},
+    Decodable, Encodable, Message, Result,
+};
+use serde::{Deserialize, Serialize};
+
+/// A message metadata wrapper type
+///
+/// This message wraps around a well-typed Message type, with
+/// additional metadata.  Metadata is split between the "scope" and
+/// "generic" sections.
+///
+/// ## Scope metadata
+///
+/// This metadata is passed around in a particular metadata scope.
+/// For example, a worker that adds some behaviour to message sending
+/// may chose to embed "scope" metadata.  When wrapping this message
+/// in another scope the previously scoped metadata becomes part of
+/// the opaque `data` section.
+///
+/// Thus it is not possible to retrieve metadata from a different
+/// nested scope!
+///
+/// ## Generic metadata
+///
+/// When creating an `OckamMessage` it's also possible to attach
+/// generic metadata.  This data is passed around for every nested
+/// scope and must be re-attached to the outest-most scope when
+/// peeling a nested message stack.
+#[derive(Message, Serialize, Deserialize)]
+pub struct OckamMessage {
+    /// Main data section of this message
+    pub data: Vec<u8>,
+    /// Metadata for this specific scope
+    pub int: Vec<Vec<u8>>,
+    /// Metadata that is carried to the final recipient of the message
+    pub ext: Option<Metadata>,
+}
+
+impl OckamMessage {
+    pub fn new<M: Message>(msg: M) -> Result<Self> {
+        Ok(Self {
+            data: msg.encode()?,
+            int: vec![],
+            ext: None,
+        })
+    }
+
+    /// Create a new `OckamMessage` by nesting a previous one
+    pub fn wrap(mut prev: Self) -> Result<Self> {
+        let ext = core::mem::replace(&mut prev.ext, None);
+        Ok(Self {
+            data: prev.encode()?,
+            int: vec![],
+            ext,
+        })
+    }
+
+    /// Add some metadata to this scope
+    pub fn scope_data(mut self, meta: Vec<u8>) -> Self {
+        self.int.push(meta);
+        self
+    }
+
+    /// Add to the generic metadata section
+    pub fn generic_data<S: Into<String>>(mut self, key: S, val: Vec<u8>) -> Self {
+        if self.ext.is_none() {
+            self.ext = Some(Metadata(BTreeMap::new()));
+        }
+
+        self.ext.as_mut().unwrap().insert(key.into(), val);
+        self
+    }
+
+    /// Get the current scope metadata
+    pub fn get_scope(&self) -> &Vec<Vec<u8>> {
```

Comment:
Nit: Given that these all are public, I'm not sure having getters is worthwhile. (And if we do want them, probably this one should be `&[Vec<u8>]` too, and they shouldn't use the `get_` prefix, since that's not generally considered good style in rust)